### ISSUE TITLE
Consolidate component stack and (DEV only) element __source

### ISSUE THREAD

**Issue by bvaughn** [facebook/react#22738]
There are currently two types of component "location" displayed and used in DevTools:
* "Native" component stacks: These are detected by intentionally causing the component to throw an error (e.g. `null` props object, a hooks dispatcher that throws) so DevTools can figure out **where the function is defined**.
* DEV source metadata that comes from plugins like [@babel/plugin-transform-react-jsx-source](https://github.com/babel/babel/blob/main/packages/babel-plugin-transform-react-jsx-source/src/index.js) and describes **where the JSX element is defined** (e.g. where its props are passed).

In other words, for the `Greeting` component:
```jsx
function Example() {
  return <Greeting name="brian" />; // <- DEV source metadata
}

function Greeting({name}) { // <- component stack
  return <div>Hi, {name}</div>;
}
```
The two locations have different use cases. Which one do you want when you click the "open in editor" button?

I'm not sure. I can imagine cases for both of these. My gut is that the component stack location is actually slightly better for "open in editor" _but_ it would be a change in behavior.

That being said, it would be nice to pick one and stick with it, and the component stack location has the added benefit of working in production.

**Comment by gaearon** [comment]
I think the plan was that maybe we get rid of `__source`. But not sure.

**Comment by bvaughn** [comment]
Yes, that would be nice. I just wanted to point out that the location of `__source` isn't something that we can emulate with our component stack approach.

I think we could emulate it by examining the stack in `ReactElement` constructor though if we really wanted to? But maybe we don't actually want to use it for anything?

**Comment by bvaughn** [comment]
> The two locations have different use cases. Which one do you want when you click the "open in editor" button?

As I think about this more, I realize a major oversight with regard to this comment.

The DEV source location describes where the _source code_ lives on the file system. The component stack location describes where the _compiled code_ lives. Opening the compiled code in VS Code is probably also not what users would want to do.
